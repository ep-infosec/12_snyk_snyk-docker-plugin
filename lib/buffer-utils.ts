import * as crypto from "crypto";
import { HashAlgorithm } from "./types";

const HASH_ENCODING = "hex";

export function bufferToSha1(buffer: Buffer): string {
  const hash = crypto.createHash(HashAlgorithm.Sha1);
  hash.setEncoding(HASH_ENCODING);
  hash.update(buffer);
  hash.end();
  return hash.read().toString(HASH_ENCODING);
}
