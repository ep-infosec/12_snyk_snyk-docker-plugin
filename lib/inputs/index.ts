import {
  FilePathToBuffer,
  FilePathToContent,
  FilePathToElfContent,
} from "../analyzer/applications/types";
import { ExtractedLayers, FileContent } from "../extractor/types";
import { Elf } from "../go-parser/types";

export function getContent(
  extractedLayers: ExtractedLayers,
  searchedAction: string,
  contentTypeValidation: (type: FileContent) => boolean,
): FilePathToContent {
  const foundAppFiles = {};

  for (const filePath of Object.keys(extractedLayers)) {
    for (const actionName of Object.keys(extractedLayers[filePath])) {
      if (actionName !== searchedAction) {
        continue;
      }
      if (!contentTypeValidation(extractedLayers[filePath][actionName])) {
        throw new Error("unexpected content type");
      }
      foundAppFiles[filePath] = extractedLayers[filePath][actionName];
    }
  }

  return foundAppFiles;
}

function isStringType(type: FileContent) {
  return typeof type === "string";
}

export function getFileContent(
  extractedLayers: ExtractedLayers,
  searchedAction: string,
): FilePathToContent {
  let foundAppFiles;
  try {
    foundAppFiles = getContent(extractedLayers, searchedAction, isStringType);
  } catch {
    throw new Error("expected string");
  }

  return foundAppFiles;
}

function isElfType(type: FileContent): type is Elf {
  const elf = type as Elf;
  return !!(elf.body && elf.body.programs && elf.body.sections);
}

export function getElfFileContent(
  extractedLayers: ExtractedLayers,
  searchedAction: string,
): FilePathToElfContent {
  let foundAppFiles;
  try {
    foundAppFiles = getContent(extractedLayers, searchedAction, isElfType);
  } catch {
    throw new Error("elf file expected to contain programs and sections");
  }

  return foundAppFiles;
}

function isTypeBuffer(type: FileContent) {
  return Buffer.isBuffer(type);
}

export function getBufferContent(
  extractedLayers: ExtractedLayers,
  searchedAction: string,
): FilePathToBuffer {
  let foundAppFiles;
  try {
    foundAppFiles = getContent(extractedLayers, searchedAction, isTypeBuffer);
  } catch {
    throw new Error("expected Buffer");
  }

  return foundAppFiles;
}
