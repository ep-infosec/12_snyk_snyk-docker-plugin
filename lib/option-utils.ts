export { isTrue, isNumber };

function isTrue(value?: boolean | string): boolean {
  return String(value).toLowerCase() === "true";
}

function isNumber(value?: boolean | string): boolean {
  return !isNaN(Number(value));
}
