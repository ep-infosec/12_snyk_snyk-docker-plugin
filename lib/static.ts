import * as analyzer from "./analyzer";
import { StaticAnalysis } from "./analyzer/types";
import { buildTree } from "./dependency-tree";
import { DockerFileAnalysis } from "./dockerfile/types";
import { isTrue } from "./option-utils";
import { parseAnalysisResults } from "./parser";
import { buildResponse } from "./response-builder";
import { DepTree, ImageType, PluginOptions, PluginResponse } from "./types";

export async function analyzeStatically(
  targetImage: string,
  dockerfileAnalysis: DockerFileAnalysis | undefined,
  imageType: ImageType,
  imagePath: string,
  globsToFind: { include: string[]; exclude: string[] },
  options: Partial<PluginOptions>,
): Promise<PluginResponse> {
  const staticAnalysis = await analyzer.analyzeStatically(
    targetImage,
    dockerfileAnalysis,
    imageType,
    imagePath,
    globsToFind,
    options,
  );

  const parsedAnalysisResult = parseAnalysisResults(
    targetImage,
    staticAnalysis,
  );

  /** @deprecated Should try to build a dependency graph instead. */
  const dependenciesTree = buildTree(
    targetImage,
    parsedAnalysisResult.packageFormat,
    parsedAnalysisResult.depInfosList,
    parsedAnalysisResult.targetOS,
  );

  const analysis: StaticAnalysis & {
    depTree: DepTree;
    packageFormat: string;
  } = {
    ...staticAnalysis,
    depTree: dependenciesTree,
    imageId: parsedAnalysisResult.imageId,
    imageLayers: parsedAnalysisResult.imageLayers,
    packageFormat: parsedAnalysisResult.packageFormat,
  };

  const excludeBaseImageVulns = isTrue(options["exclude-base-image-vulns"]);
  return buildResponse(analysis, dockerfileAnalysis, excludeBaseImageVulns);
}
