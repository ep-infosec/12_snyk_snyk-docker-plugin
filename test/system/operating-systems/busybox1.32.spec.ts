import { scan } from "../../../lib/index";
import { execute } from "../../../lib/sub-process";

describe("busybox tests", () => {
  afterAll(async () => {
    await execute("docker", [
      "image",
      "rm",
      "busybox@sha256:32dc7202280d641c9c0d94a13247195adc83dba83aca0eed383888c01311cfc2",
    ]).catch();
  });

  it("should correctly analyze a busybox image by sha256", async () => {
    const image =
      "busybox@sha256:32dc7202280d641c9c0d94a13247195adc83dba83aca0eed383888c01311cfc2";
    const pluginResult = await scan({
      path: image,
    });

    expect(pluginResult).toMatchSnapshot();
  });
});
