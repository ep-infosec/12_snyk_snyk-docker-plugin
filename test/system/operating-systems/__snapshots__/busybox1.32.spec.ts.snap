// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`busybox tests should correctly analyze a busybox image by sha256 1`] = `
Object {
  "scanResults": Array [
    Object {
      "facts": Array [
        Object {
          "data": Object {
            "graph": Object {
              "nodes": Array [
                Object {
                  "deps": Array [],
                  "nodeId": "root-node",
                  "pkgId": "docker-image|busybox@",
                },
              ],
              "rootNodeId": "root-node",
            },
            "pkgManager": Object {
              "name": "linux",
              "repositories": Array [
                Object {
                  "alias": "unknown:0.0",
                },
              ],
            },
            "pkgs": Array [
              Object {
                "id": "docker-image|busybox@",
                "info": Object {
                  "name": "docker-image|busybox",
                  "version": undefined,
                },
              },
            ],
            "schemaVersion": "1.3.0",
          },
          "type": "depGraph",
        },
        Object {
          "data": "sha256:9f00140cde00e263233ec6214cd1d8b8a0f9f92c1834d5dc842d2e184e807b29",
          "type": "imageId",
        },
        Object {
          "data": Array [
            "d24325657bb21b922a4cd584d4d665c24617cb4137f0c41967e6db61ba878ef4/layer.tar",
          ],
          "type": "imageLayers",
        },
        Object {
          "data": "2020-09-09T01:38:21.750425221Z",
          "type": "imageCreationTime",
        },
        Object {
          "data": Array [
            "sha256:e8e544649a797bb64853d4b70feaea366aedb53658277861b685dd0a3e76343a",
          ],
          "type": "rootFs",
        },
      ],
      "identity": Object {
        "args": Object {
          "platform": "linux/amd64",
        },
        "type": "linux",
      },
      "target": Object {
        "image": "docker-image|busybox",
      },
    },
  ],
}
`;
